File: Source/Other/DatabaseAbstraction/SPPostgreSQLStreamingResultWrapper.m
//
//  SPPostgreSQLStreamingResultWrapper.m
//  Sequel Ace
//

#import "SPPostgreSQLStreamingResultWrapper.h"

@implementation SPPostgreSQLStreamingResultWrapper {
    SPPostgreSQLStreamingResult *_result;
    NSArray<NSString *> *_fieldNames;
    int _numFields;
    double _queryTime;
    unsigned long long _totalRowsFetched;

    // Current in-memory batch state
    int _currentBatchSize;
    int _batchRowIndex;
}

- (instancetype)initWithStreamingResult:(SPPostgreSQLStreamingResult *)result
                              queryTime:(double)queryTime {
    if (self = [super init]) {
        _result = result;
        _queryTime = queryTime;
        _numFields = sp_postgresql_streaming_result_num_fields(result);
        _totalRowsFetched = 0;
        _currentBatchSize = 0;
        _batchRowIndex = 0;

        NSMutableArray *names = [NSMutableArray arrayWithCapacity:_numFields];
        for (int i = 0; i < _numFields; i++) {
            char *name = sp_postgresql_streaming_result_field_name(result, i);
            if (name) {
                [names addObject:[NSString stringWithUTF8String:name]];
                sp_postgresql_free_string(name);
            } else {
                [names addObject:[NSString stringWithFormat:@"col%d", i]];
            }
        }
        _fieldNames = [names copy];

        // Prime the first batch
        [self _fetchNextBatch];
    }
    return self;
}

- (void)dealloc {
    if (_result) {
        sp_postgresql_streaming_result_destroy(_result);
        _result = NULL;
    }
}

- (void)markConnectionDisconnected {
    if (_result) {
        sp_postgresql_streaming_result_mark_disconnected(_result);
    }
}

// ─── Batch management ─────────────────────────────────────────────────────────

- (void)_fetchNextBatch {
    if (!_result || !sp_postgresql_streaming_result_has_more(_result)) {
        _currentBatchSize = 0;
        return;
    }
    // The callback is unused in our API (values fetched via get_batch_value)
    extern int sp_postgresql_streaming_result_next_batch(SPPostgreSQLStreamingResult*, void*, void*);
    int rows = sp_postgresql_streaming_result_next_batch(_result, NULL, NULL);
    _currentBatchSize = MAX(rows, 0);
    _batchRowIndex = 0;
}

// ─── SPDatabaseResult ────────────────────────────────────────────────────────

- (NSUInteger)numberOfFields { return (NSUInteger)_numFields; }

- (unsigned long long)numberOfRows {
    // For streaming results, total rows are unknown until exhausted.
    long long total = _result ? sp_postgresql_streaming_result_total_rows(_result) : 0;
    return total >= 0 ? (unsigned long long)total : _totalRowsFetched;
}

- (NSArray *)fieldNames    { return _fieldNames; }
- (double)queryExecutionTime { return _queryTime; }
- (void)seekToRow:(unsigned long long)row { /* Streaming results cannot seek backwards. */ }

- (NSArray *)getRowAsArray {
    if (!_result) return nil;

    // Advance to the next batch when we run out of rows in the current one.
    while (_batchRowIndex >= _currentBatchSize) {
        if (!sp_postgresql_streaming_result_has_more(_result)) return nil;
        [self _fetchNextBatch];
        if (_currentBatchSize == 0) return nil;
    }

    int batchRow = _batchRowIndex++;
    _totalRowsFetched++;

    NSMutableArray *arr = [NSMutableArray arrayWithCapacity:_numFields];
    for (int col = 0; col < _numFields; col++) {
        char *val = sp_postgresql_streaming_result_get_batch_value(_result, batchRow, col);
        if (val) {
            [arr addObject:[NSString stringWithUTF8String:val]];
            sp_postgresql_free_string(val);
        } else {
            [arr addObject:[NSNull null]];
        }
    }
    return [arr copy];
}

- (NSDictionary *)getRowAsDictionary {
    NSArray *row = [self getRowAsArray];
    if (!row) return nil;
    NSMutableDictionary *dict = [NSMutableDictionary dictionaryWithCapacity:_numFields];
    for (NSUInteger i = 0; i < row.count && i < _fieldNames.count; i++) {
        dict[_fieldNames[i]] = row[i];
    }
    return [dict copy];
}

- (id)getRow { return [self getRowAsArray]; }

// NSFastEnumeration
- (NSUInteger)countByEnumeratingWithState:(NSFastEnumerationState *)state
                                  objects:(id __unsafe_unretained [])buffer
                                    count:(NSUInteger)len {
    if (state->state == 0) {
        state->mutationsPtr = &state->extra[0];
        state->state = 1;
    }
    NSUInteger count = 0;
    while (count < len) {
        NSArray *row = [self getRowAsArray];
        if (!row) break;
        buffer[count++] = row;
    }
    state->itemsPtr = buffer;
    return count;
}

@end

File: Source/Other/DatabaseAbstraction/SPPostgreSQLStreamingResultWrapper.swift
//
//  SPPostgreSQLStreamingResultWrapper.swift
//  Sequel Ace
//

import Foundation

@objc(SPPostgreSQLStreamingResultWrapper)
final class SPPostgreSQLStreamingResultWrapper: NSObject, SPDatabaseResult {
    private var result: OpaquePointer?
    private let storedFieldNames: [String]
    private let numFields: Int32
    private let queryTime: Double
    private var totalRowsFetched: UInt64 = 0
    private var currentBatchSize: Int32 = 0
    private var batchRowIndex: Int32 = 0

    @objc(initWithStreamingResult:queryTime:)
    init(streamingResult: OpaquePointer, queryTime: Double) {
        self.result = streamingResult
        self.queryTime = queryTime
        self.numFields = sp_postgresql_streaming_result_num_fields(streamingResult)

        var names = [String]()
        for i in 0..<numFields {
            if let namePtr = sp_postgresql_streaming_result_field_name(streamingResult, i) {
                names.append(String(cString: namePtr))
                sp_postgresql_free_string(namePtr)
            } else {
                names.append("col\(i)")
            }
        }
        self.storedFieldNames = names
        super.init()
        fetchNextBatch()
    }

    deinit {
        if let result {
            sp_postgresql_streaming_result_destroy(result)
            self.result = nil
        }
    }

    @objc func markConnectionDisconnected() {
        if let result {
            sp_postgresql_streaming_result_mark_disconnected(result)
        }
    }

    private func fetchNextBatch() {
        guard let result, sp_postgresql_streaming_result_has_more(result) != 0 else {
            currentBatchSize = 0
            return
        }
        let rows = sp_postgresql_streaming_result_next_batch(result, nil, nil)
        currentBatchSize = max(rows, 0)
        batchRowIndex = 0
    }

    func numberOfFields() -> UInt { UInt(numFields) }

    func numberOfRows() -> UInt64 {
        guard let result else { return totalRowsFetched }
        let total = sp_postgresql_streaming_result_total_rows(result)
        return total >= 0 ? UInt64(total) : totalRowsFetched
    }

    func fieldNames() -> [String] { storedFieldNames }
    func queryExecutionTime() -> Double { queryTime }
    func seekToRow(_ row: UInt64) { /* streaming results cannot seek backwards */ }

    func getRowAsArray() -> [Any]? {
        guard let result else { return nil }

        while batchRowIndex >= currentBatchSize {
            guard sp_postgresql_streaming_result_has_more(result) != 0 else { return nil }
            fetchNextBatch()
            if currentBatchSize == 0 { return nil }
        }

        let batchRow = batchRowIndex
        batchRowIndex += 1
        totalRowsFetched += 1

        var arr = [Any]()
        arr.reserveCapacity(Int(numFields))
        for col in 0..<numFields {
            if let valPtr = sp_postgresql_streaming_result_get_batch_value(result, batchRow, col) {
                arr.append(String(cString: valPtr))
                sp_postgresql_free_string(valPtr)
            } else {
                arr.append(NSNull())
            }
        }
        return arr
    }

    func getRowAsDictionary() -> [AnyHashable: Any]? {
        guard let row = getRowAsArray() else { return nil }
        var dict = [AnyHashable: Any]()
        for (index, value) in row.enumerated() where index < storedFieldNames.count {
            dict[storedFieldNames[index]] = value
        }
        return dict
    }

    func getRow() -> Any? { getRowAsArray() }

    @objc(countByEnumeratingWithState:objects:count:)
    func countByEnumerating(with state: UnsafeMutablePointer<NSFastEnumerationState>,
                            objects buffer: AutoreleasingUnsafeMutablePointer<AnyObject?>,
                            count len: Int) -> Int {
        if state.pointee.state == 0 {
            state.pointee.mutationsPtr = withUnsafeMutablePointer(to: &state.pointee.extra.0) { $0 }
            state.pointee.state = 1
        }
        var count = 0
        while count < len {
            guard let row = getRowAsArray() else { break }
            buffer.advanced(by: count).pointee = row as NSArray
            count += 1
        }
        state.pointee.itemsPtr = buffer
        return count
    }
}
